### PR TITLE
Add --dump-core option

### PR DESCRIPTION
**Pull request by amosr**
This might not be the best way to go about it. Just an idea

### CODE REVIEW

**Comment by amosr** [comment]
@jystic @charleso

**Comment by jacobstanley** [comment]
this is cool, will take a proper look tomorrow

**Comment by charleso** [comment]
I really know nothing about the flags in question.

Just curious, is it already possible to pass ghc flags though? I guess I'm asking - does this need to be in mafia or is it just a call-site decision?

**Comment by HuwCampbell** [comment]
One can use arguments which are passed to cabal after the `--`, and `--ghc-options=*` will shuffle them there.

So I think this is for convenience as there's 4 flags required here.

**Comment by jacobstanley** [comment]
Yeah it's definitely convenience, I for one can never remember what the flags are called.

I think maybe `-dsuppress-all` is a little too aggressive, I quite like having the type signatures in the output or I find it harder to decipher what's going on. Would you be ok with using `-dsuppress-coercions -dsuppress-idinfo -dsuppress-module-prefixes` in place of `-dsuppress-all`? It's basically the same but with type signatures.

What is the difference between between `-ddump-prep` and `-ddump-simpl`? For some reason `-ddump-prep` always gives me a file with two dumps right after another which appear to be identical?

I think I would prefer to use `-ddump-simpl` or maybe we should have two options in mafia `--dump-simpl` and `--dump-prep` (instead of `--dump-core`), what do you think? /not sure

**Comment by jacobstanley** [comment]
> maybe we should have two options in mafia --dump-simpl and --dump-prep (instead of --dump-core), what do you think? /not sure

Having said this, I do like the simplicity of having only a single `--dump-core` option. Part of the benefit of this feature is that it makes it trivial to get started inspecting core, even if you've never done it before.

**Comment by tranma** [comment]
I'm 👍 for `-ddump-core`

can also use something from this https://github.com/shachaf/ghc-core/blob/master/ghc-core.hs to make it more readable, if necessary

**Comment by HuwCampbell** [comment]
It's similar to the profiling flag, in that there's scope in mafia for making things which ghc and cabal make hard easy.

File: main/mafia.hs [hunk 1]
```diff
@@ -344,7 +355,15 @@ mafiaBuild p w flags args = do
         EnableWarnings ->
           ["--ghc-options=-Werror"]
 
-  liftCabal . cabal_ "build" $ ["-j"] <> wargs <> args
+    dumpargs =
+      case dump of
+        DisableCoreDump ->
+          []
+        EnableCoreDump ->
+          ["--ghc-options=-ddump-prep -ddump-to-file -dppr-case-as-let -dsuppress-all"]
```
Comment (jacobstanley) [inline main/mafia.hs]: You can mention `--ghc-options` multiple times, so I think I would prefer this to be:

``` hs
[ "--ghc-options=-ddump-prep"
, "--ghc-options=-ddump-to-file"
, "--ghc-options=-dppr-case-as-let"
-- etc --
]
```

Comment (amosr) [reply main/mafia.hs]: Yeah fair enough, I'm not too worried either way, will change

**Comment by amosr** [comment]
Yeah, it really is just a shorthand. Maybe we want a more principled way of doing shortcuts that expand to multiple options?

**Comment by amosr** [comment]
> What is the difference between between -ddump-prep and -ddump-simpl? For some reason -ddump-prep always gives me a file with two dumps right after another which appear to be identical?

It's just cargo-cult muscle memory for me. I think prep is slightly after simpl in the pipeline, but I don't think it makes a real difference. Happy to go with simpl

**Comment by amosr** [comment]
> can also use something from this https://github.com/shachaf/ghc-core/blob/master/ghc-core.hs to make it more readable, if necessary

I just had a look, maybe those regexes are superseded by the -dsuppress options?

**Comment by jacobstanley** [comment]
:wolf: so good

**Comment by erikd-ambiata** [comment]
> I just had a look, maybe those regexes are superseded by the -dsuppress options?

@amosr That might be the case. If I remember correctly ghc-core has other features beyond just that. Its a pity its been un-maintained for so long.

**Pull request merged**